File: android/pi-tools/src/main/kotlin/com/pimobile/tools/WriteFileTool.kt
package com.pimobile.tools

import com.pimobile.agent.AgentToolResult
import com.pimobile.agent.Tool
import kotlinx.serialization.json.*
import java.io.File

class WriteFileTool(private val sandboxDir: File) : Tool {

    override val name = "write_file"
    override val description = "Write content to a file in the sandbox directory. Creates parent directories if needed."
    override val parametersSchema = buildJsonObject {
        put("type", "object")
        putJsonObject("properties") {
            putJsonObject("path") {
                put("type", "string")
                put("description", "File path relative to sandbox directory")
            }
            putJsonObject("content") {
                put("type", "string")
                put("description", "Content to write to the file")
            }
        }
        putJsonArray("required") { add("path"); add("content") }
    }

    override suspend fun execute(input: JsonObject): AgentToolResult {
        val path = input["path"]?.jsonPrimitive?.contentOrNull
            ?: return AgentToolResult("", "Error: 'path' is required", isError = true)
        val content = input["content"]?.jsonPrimitive?.contentOrNull
            ?: return AgentToolResult("", "Error: 'content' is required", isError = true)

        return try {
            val file = File(sandboxDir, path).canonicalFile
            if (!file.canonicalPath.startsWith(sandboxDir.canonicalPath)) {
                return AgentToolResult("", "Error: Path is outside sandbox", isError = true)
            }

            file.parentFile?.mkdirs()
            file.writeText(content)

            val bytes = file.length()
            AgentToolResult("", "Successfully wrote $bytes bytes to $path")
        } catch (e: Exception) {
            AgentToolResult("", "Error writing file: ${e.message}", isError = true)
        }
    }
}

File: ios/PiTools/Sources/PiTools/WriteFileTool.swift
import Foundation
import PiAgentCore
import PiAI

public struct WriteFileTool: Tool, Sendable {
    public let name = "write_file"
    public let description = "Write content to a file. Creates parent directories if needed. Restricted to the sandbox directory."

    private let sandboxURL: URL

    public init(sandboxURL: URL) {
        self.sandboxURL = sandboxURL
    }

    public var parametersSchema: JSONValue {
        .object([
            "type": .string("object"),
            "required": .array([.string("path"), .string("content")]),
            "properties": .object([
                "path": .object([
                    "type": .string("string"),
                    "description": .string("Path to the file to write, relative to sandbox"),
                ]),
                "content": .object([
                    "type": .string("string"),
                    "description": .string("Content to write to the file"),
                ]),
            ]),
        ])
    }

    public func execute(input: JSONValue) async throws -> AgentToolResult {
        guard let path = input["path"]?.stringValue else {
            return AgentToolResult(output: "Error: 'path' parameter is required", isError: true)
        }
        guard let content = input["content"]?.stringValue else {
            return AgentToolResult(output: "Error: 'content' parameter is required", isError: true)
        }

        let fileURL = sandboxURL.appendingPathComponent(path)

        // Validate path is within sandbox
        guard fileURL.standardizedFileURL.path.hasPrefix(sandboxURL.standardizedFileURL.path) else {
            return AgentToolResult(output: "Error: Path is outside sandbox", isError: true)
        }

        let fm = FileManager.default
        let directory = fileURL.deletingLastPathComponent()

        // Create parent directories
        if !fm.fileExists(atPath: directory.path) {
            do {
                try fm.createDirectory(at: directory, withIntermediateDirectories: true)
            } catch {
                return AgentToolResult(output: "Error creating directory: \(error.localizedDescription)", isError: true)
            }
        }

        // Write file
        guard let data = content.data(using: .utf8) else {
            return AgentToolResult(output: "Error: Could not encode content as UTF-8", isError: true)
        }

        do {
            try data.write(to: fileURL)
        } catch {
            return AgentToolResult(output: "Error writing file: \(error.localizedDescription)", isError: true)
        }

        let lineCount = content.components(separatedBy: "\n").count
        return AgentToolResult(
            output: "Wrote \(data.count) bytes (\(lineCount) lines) to \(path)",
            details: .file(path: path, content: content, language: nil)
        )
    }
}
